### ISSUE TITLE
container should poll until mysql is alive or until timeout occurs

### ISSUE THREAD

**Issue by BelfordZ** [flyway/flyway-docker#2]
As mentioned in this issue: https://github.com/dhoer/docker-flyway/issues/4

I would love for this docker image to use a wait-for script to poll for mysql before running the commands. 

thoughts? Id be happy to make a PR for this.

**Comment by axelfontaine** [comment]
The Flyway container isn't the place for this. Instead the MySQL Docker images should implement a [HEALTHCHECK](https://docs.docker.com/engine/reference/builder/#healthcheck). Better and more accurate that any kind of wait-for scripts in dependent containers. Please raise an issue with them.

**Comment by BelfordZ** [comment]
Roger that @axelfontaine thanks for the help!! Thats a much more elegant solution. I wasn't aware that was a feature in docker!

**Comment by larryboymi** [comment]
Part of the issue is that flyway needs the database to be up, but docker relies on the containers to do their own healthchecks to sort out dependencies. Having the ability to verify a login is a good thing. My guess is that you could probably have the healthcheck be a `flyway info` command where you checked the return code after logging in.

**Comment by BelfordZ** [comment]
@larryboymi Here is the solution that worked for me. Note that it does not work on latest docker as depends_on condition doesnt work the same anymore. THis is why I set the version to 2.3
```
version: "2.3"
services:
  schema:
    image: dhoer/flyway:mysql
    command: -url=jdbc:mysql://mysql -schemas=MYSUPERCOOLDATABASE -user=root migrate
    volumes:
      - .:/flyway/sql
    depends_on:
      mysql:
        condition: service_healthy

  mysql:
    image: mysql:latest
    ports:
      - "3306:3306"
    environment:
      MYSQL_ALLOW_EMPTY_PASSWORD: "yes"
      MYSQL_DATABASE: "MYSUPERCOOLDATABASE"
    healthcheck:
      test: mysql -e "SHOW VARIABLES" -u root >/dev/null 2>&1;
      timeout: 20s
      retries: 10
```

Hope this helps

**Comment by larryboymi** [comment]
@BelfordZ thanks, but yeah, we're using the latest compose/swarm/stack, and I think healthchecks are all that is left.

**Comment by BelfordZ** [comment]
Yeah. Correct me if I'm wrong, but it seems the current direction docker is heading in is: deal with waiting for dependent services to be alive inside your own container, don't rely on the healthchecks for this. Most people are using some variant of [waits-for.sh](https://github.com/blueimp/shell-scripts/blob/master/bin/wait-for.sh), but in the case of flyway, there is an entrypoint set, and so setting a commad that runs a waits-for shell script doesnt work.

If thats the case, my recommendation would be to reopen this issue.

Thoughts @larryboymi @axelfontaine ?

**Comment by BelfordZ** [comment]
further, mysql folks are 100% never going to implement a healthcheck, as the concept of 'healthy' when it comes to mysql is not straightforward in all cases.

**Comment by larryboymi** [comment]
I was thinking of combining the compose healthcheck's `start_period` with a healthcheck doing a `flyway info` on the DB (not sure what that returns if flyway has never run), and hoping that returned a bad code if it couldn't connect to the database. Then setting the delay/retries to a reasonable amount.

**Comment by BelfordZ** [comment]
AH that would be nice. AFAIK the flyway container's migrate command wouldn't wait for the healthcheck to pass before running though, so I'm not sure it would solve the issue.

**Comment by larryboymi** [comment]
@BelfordZ My hope is that a healthcheck has to pass in order for the container to be "started" (or run it's `CMD`)

**Comment by jvitor83** [comment]
@BelfordZ @larryboymi 

You can override the entrypoint with `sh` and then use the command to exec a sleep and then flyway.
```
version: '3'
services:

  flyway:
    image: boxfuse/flyway
    command: -c "sleep 30; flyway migrate -user=sa -password=P@ssw0rd -url=jdbc:sqlserver://sqlserver:1433;databaseName=tempdb"
    entrypoint: sh
    volumes:
      - ./sql:/flyway/sql

  sqlserver:
    image: microsoft/mssql-server-linux:2017-latest
    ports:
      - 1433:1433
    environment:
      ACCEPT_EULA: 'Y'
      SA_PASSWORD: P@ssw0rd
```
it works for me

**Comment by ido-namely** [comment]
I also think this issue should be addressed by flyway itself,and for all DB's not just MySQL.
Since docker compose no longer support health checks, we know users are going to implement their own sleep/wait, why not integrate it in the flyway image itself, if for whatever reason anyone doesn't want/like it, we can set it as an argument.
I'll be happy to submit a PR for this if you agree

**Comment by axelfontaine** [comment]
@ido-namely No true. See https://docs.docker.com/compose/compose-file/#healthcheck

That being said, introducing something like a connect timeout within Flyway itself could also be an option. Currently Flyway automatically tries to connect up to 10 times at a 1 second interval. This could be made configurable and it would serve all users, not just the Docker ones.

**Comment by ido-namely** [comment]
@axelfontaine I misspoke, I meant the healthcheck is not longer addressed as a condition for the `depends_on` option https://docs.docker.com/compose/compose-file/#depends_on since v. 3

So when setting flyway container as a dependency on the DB container , you would probably need to add a `wait for` script or add `sleep` as suggested in this thread.
I feel it would be very beneficial for users of the flyway container and docker-compose to have this option built in.

**Comment by axelfontaine** [comment]
This has now been implemented and will be released as part of Flyway 5.2.0.

The Flyway container will then support the `FLYWAY_CONNECT_RETRIES` environment variable to specify the maximum number of retries to perform at a 1 second interval when attempting to connect to the database. For example, setting this to `3600` will make the Flyway container try to connect for up to 1 hour.

This works with all supported databases and doesn't impose an HEALTHCHECK requirements on the database container.
